invenio_records_resources/links.py

```diff
@@ -135,3 +119,113 @@ def links(self):
             _links["next"] = next_link
 
         return _links
+
+
+class Linker:
+    """Linker class.
+
+    Generates all the links for a given namespace.
+    """
+
+    def __init__(self, link_builders):
+        """Constructor.
+
+        :param link_builders: dict(string, list<LinkBuilder>)
+        """
+        self.link_builders = link_builders
+
+    def links(self, namespace, identity, **kwargs):
+        """Returns dict of links."""
+        output_links = {}
+        for link_builder in self.link_builders[namespace]:
+            link = link_builder.route_to_link(identity, **kwargs)
+            output_links.update(link)
+        return output_links
+
+    def register_link_builders(self, link_builders):
+        """Updates the internal link_builders with new ones."""
+        self.link_builders.update(link_builders)
+
+
+class RecordLinkBuilder:
+    """Common interface among most record link builders."""
+
+    def __init__(self, key, route, action, permission_policy):
+        """Constructor."""
+        self.key = key
+        self.route = route
+        self.action = action
+        self.permission_policy = permission_policy
+
+    def route_to_link(self, identity, **kwargs):
+        """Converts route to a link."""
+        if self.permission_policy(self.action, **kwargs).allows(identity):
+            pid_value = kwargs["pid_value"]
+            path = self.route.replace("<pid_value>", pid_value)
+            return {self.key: base_url(path=path)}
+        else:
+            return {}
+
+
+class RecordSelfLinkBuilder(RecordLinkBuilder):
+    """Builds record "self" link."""
+
+    def __init__(self, config):
+        """Constructor."""
+        super(RecordSelfLinkBuilder, self).__init__(
```

Comment:
normal: There are four classes just overwriting the constructor. Could they just be instances?

```python
record_self_link_builder = RecordLinkBuilder(key="self",
             route=api_route(config.record_route),
             action="read",
             permission_policy=config.permission_policy_cls
         )
```

Same for the few below. I think there is no need to create classes here. In the end if someone needs a change, they will create another instance.

Reply:
The `config` is dynamically provided at application runtime, so we can't assign it beforehand. Writing this in-place is a bit too much as well. (and finally I am not a fan of global assignments if I can do it another way :cat:  )